Engine code and agent code

anikde opened the pull request:
3 files added

anikde committed:
engine code and agent code

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+		if (self.gamestate[a] != 0):
+			return Code.FAILURE
+		else: 
+			return Code.SUCCESS
```
Please add a new line at EOF

anikde replied:
ok done

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+from enum import Enum
+
+state_list=[]
+class Code(Enum):
```
Consider adding `@unique` to enum. Ref: https://docs.python.org/3/library/enum.html#enum.unique

`Code` is a very generic name which can lead to ambiguity:
- It serves as a _return code_
- These return codes are exclusive to a particular class `tictactoeEngine` and should **not** be used to infer return values from any other class.
  - Maybe consider making the enum class as **inner class** of tictactoeEngine. ref: https://www.datacamp.com/community/tutorials/inner-classes-python

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,21 @@
+		json.dump(returnn[2], json_file, indent="\t")
+		json_file.write("\t")
+		json.dump(list(returnn[1]), json_file)
+		json_file.write("\n")
```
dump / file write should happen only once... certainly not in loop... 
dump can do file write. ref: https://realpython.com/python-json/#a-simple-serialization-example
use numerical value in `indent` rather than `"\t"`

Steps:
- all the required data should be first stored as a python dictionary (within a loop in this example)
- once all required data has been stored in a python dictionary, convert and write the dictionary into a json file. use the example given in the link above as a reference.


Py + JSON refs:
- Simplest example: https://docs.python-guide.org/scenarios/json/
- Detailed: https://www.freecodecamp.org/news/python-read-json-file-how-to-load-json-from-a-file-and-parse-dumps/

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+	VALID = 6
+
+class tictactoeEngine:
+	def __init__(self, agent1, agent1moves, agent1value,  agent2, agent2moves, agent2value):
```
This requires that both players should provide all sequence of moves before the game can start. That is unrealistic.
How can `player1` make a second move without knowing what was the first move of `player2`??

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+		self.index1 = 0
+		self.index2 = 0
+
+	def startEngine(self):
+		if (self.gamestate == [0] * 9):
+			if (self.checkWin() == Code.RUNNING):
+				for x in range(0,9):
+					if(x%2 == 0 and (self.checkTie() != Code.TIE) and (self.checkWin()== Code.RUNNING) ):
+						a = self.agent1moves[self.index1]
+						while(self.checkvalid(a-1) != Code.SUCCESS):
+							a = self.agent1moves[self.index1]
+						self.index1 +=1
+						self.gamestate[a-1] = self.agent1value
+					elif(x%2==1 and (self.checkTie() != Code.TIE) and (self.checkWin()== Code.RUNNING) ):
+						b = self.agent2moves[self.index2]
+						while(self.checkvalid(b-1) != Code.SUCCESS):
+							b = self.agent2moves[self.index2]
+						self.index2 +=1
+						self.gamestate[b-1] = self.agent2value
+
+					self.get_gameState()
+					if (Code.TIE == self.checkTie()): 
+						return Code.TIE, 0, 0
+						 
+					if (self.checkWin()== Code.WIN and (self.checkTie() != Code.TIE) ):
+						if (x%2==0): 
+							return Code.SUCCESS , self.get_gameState(), self.agent1value  
+						else:
+							return Code.SUCCESS, self.get_gameState(), self.agent2value
```
Why does such a complicated function exist in game engine?
Where is the function to make a single move? How can a player / agent / interface get the current board and make a move?

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+from enum import Enum
+
+state_list=[]
```
There should not be a global variable. The states can be a member of the game engine or interface.

anikde replied:
yes this can be done, I did that first but wasn't working fine
Now it's working

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,91 @@
+	
+
+	def checkTie(self):
+		count = 0 
+		for p in self.gamestate:
+			if (p != 0):
+				count +=1 
+		if (count==9): return Code.TIE
+		return Code.RUNNING
```
This code can be done more efficiently without local variable count.
If any of the game state is 0 then return RUNNING. else by default return TIE.

anikde replied:
Didn't get that completely

deysumitkr replied:
You do not need the local variable `count` for achieving the same functionality:
```
for p in self.gamestate:
   if (p == 0):
       return Code.RUNNING
return return Code.TIE
```

if you want to be bit more fancy:
```
if all(self.gamestate):
    return Code.TIE
else:
    Code.RUNNING
```
which can be further turned into one line code:
```
return Code.TIE if all(self.gamestate) else Code.RUNNING
```

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,21 @@
+# def test1():
+moves1 = [1, 2, 6, 4, 9]
+moves2 = [7, 5, 3, 8]
+agent1 = agent("anik", moves1, 1)
+agent1_ready = agent1.agent_instance()
+agent2 = agent("arka", moves2, -1)
+agent2_ready = agent2.agent_instance()
+engine_input = tictactoeEngine(agent1_ready[0], agent1_ready[1], agent1_ready[2], agent2_ready[0], agent2_ready[1], agent2_ready[2] ) 
```
Class variables can be directly accessed by their objects. (examples: `agent1.name`, `agent2.moves` etc)
Functions like `agent_instance()` are not required to access the member variables.

deysumitkr reviewed (changes requested):
The design of `tictactoeEngine` does not fulfill the minimum requirements:
- the players should not be required to list the sequence of moves blindly before the start of the game.
- it should be possible to play a single move at a time.
- it should be possible to get a list of available moves before playing a single move.

---

The design of `agent` does not fulfill the minimum requirements:
- it should be possible to ask agent for a move based on a given board state and available moves.
  - the current agent should choose a random move from the list of available moves. 
  - consider `random.choice` to draw a sample from a list. ref: https://docs.python.org/3/library/random.html#random.choice

deysumitkr commented:
What should `runengine.py` do?

Play one game:

- initialize two agents
- initialize a gameEngine
- initialize data-structure to store: board_state, player_moves etc.
- start loop: while game is RUNNING
  - get current board from gameEngine
  - get available moves from the gameEngine
  - get a move from concerned agent by providing current board and available moves
  - update the gameEngine with the move
  - store the board_state, player in the data-structure
- if the gameEngine resulted in win/loss status:
  - dump the data-structure into a json file

(`runengine.py` acts as an interface here)

---

In case, a test is desired then it should be a separate file on its own.
References:
- https://realpython.com/python-testing/
- https://code.visualstudio.com/docs/python/testing

anikde committed:
Changes done to (agent_class.py) --> agent_random.py
	- added a function move which takes the current board state and randomly chooses from the vacant places
changes done to tictactoe_engine.py
	- enumeration was made unique
	- engine can now ask the agent for input providing it with current game state
changes done to runengine.py
	- two agents can be initialized which are then passed to the engine
	- dictionary was created to store the winning sates and the winner's identity
	- json dump was used to dump store the dictionary in json format

anikde committed:
removed commented lines

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -10,82 +11,77 @@ class Code(Enum):
+						b = self.agent2.move(self.gamestate)
+						while(self.checkvalid(b) != engineReturnCode.SUCCESS):
+							b = (self.agent2).move(self.gamestate)
+						self.gamestate[b] = self.agent2.value
```
- Game engine should not be controlling the agents.
- Game engine should not be affected/bothered by what functions agent has and how those functions should be called to make it play. _Implies: no calls to agent or its functions from game engine._

If it was not clear already, then _this function `startEngine` should not exist._ 
This game engine is an environment (not a controller) - it does not make anyone act. It only responds to actions on it. Accept an action, if action is valid then change the state else report the problem.
The action can come from agent, human etc. - source of action should not matter.

There should be two (simpler) functions instead:
For an agent / human / interface
- to get current board
- to play a move

_Carry over from last review:_

> Where is the function to make a single move? How can a player / agent / interface get the current board and make a move?

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,24 @@
+
+if __name__ == "__main__" :
+	for x in range(11):
+		game(x)
```
New line at EOF missing (check VS code settings sent over rocket chat to make this automatic)

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,24 @@
+		"winner_moves" : returnn[1],
+		"winning_agent_value" : returnn[2]
+	}
+	if (returnn[2] != 0):
```
Such checks with constant values are confusing. Therefore, enums should be used in place to make it readable.

deysumitkr reviewed (changes requested):
`runengine.py` should be the interface that connects agents to the game engine. There should _not_ be any direct connection between agent and engine.

Please refer to what `runengine.py` should do, and from that, deduce what should be present in `tictactoe_engine.py` and `agent.py` to make that work.

Copy from previous review:
> What should runengine.py do?
> 
> Play one game:
> 
> - initialize two agents
> - initialize a gameEngine
> - initialize data-structure to store: board_state, player_moves etc.
> - start loop: while game is RUNNING
>   - get current board from gameEngine
>   - get available moves from the gameEngine
>   - get a move from concerned agent by providing current board and available moves
>   - update the gameEngine with the move
>   - store the board_state, player in the data-structure
> - if the gameEngine resulted in win/loss status:
>   - dump the data-structure into a json file

anikde committed:
Engine now works a separate entity.
Agent also works a separate entity now.
runengine works as a interface to both the classes (tictactoeEngine and agent)

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -3,85 +3,64 @@
-		return engineReturnCode.RUNNING
+		if all (self.gamestate):
+			return engineReturnCode.TIE
+		else: return engineReturnCode.RUNNING
```
why `else: return` instead of only `return`?

deysumitkr commented on agent_random.py:
```diff
@@ -0,0 +1,16 @@
+
+	def move(self, gamestate):
+		self.gamestate = gamestate
+		list =[]
```
`list` is a type name. please name the variable something else.

deysumitkr commented on agent_random.py:
```diff
@@ -0,0 +1,16 @@
+		list =[]
+		if(self.gamestate == [0]*9): return random.randint(0, 8)
+		for x in range(len(self.gamestate)):
+			if (self.gamestate[x] == 0): (list).append(x)
```
Ideally, possible actions should be obtained from environment and should not be evaluated by agent

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,38 @@
+
+agent1 = agent("anik", 1)
+agent2 = agent("arka", 5)
+engine1 = tictactoeEngine()
```
why name `engine`**1**?

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,66 @@
+			elif( self.checkstatus() == engineReturnCode.RUNNING ):
+				self.engine_status =  engineReturnCode.RUNNING
+				self.winning_agents_input = 0
+		self.get_gameState()
```
This function should return if the update was success or otherwise. That is why we have return codes.

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,66 @@
+		self.list_of_states.append(list(self.gamestate))
+		return self.list_of_states
+
+	def checkvalid(self, a):
```
- please provide a meaningful name for the variable instead of naming it as `a`
- this function can return a return code

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,66 @@
+
+	def checkWin(self):
+		if (self.gamestate[0]== self.gamestate[4] ==  self.gamestate[8] != 0 ):return engineReturnCode.WIN
+		elif(self.gamestate[0]== self.gamestate[3] ==  self.gamestate[6] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[0]== self.gamestate[1] ==  self.gamestate[2] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[2]== self.gamestate[5] ==  self.gamestate[8] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[2]== self.gamestate[4] ==  self.gamestate[6] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[6]== self.gamestate[7] ==  self.gamestate[8] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[3]== self.gamestate[4] ==  self.gamestate[5] != 0):return engineReturnCode.WIN
+		elif(self.gamestate[1]== self.gamestate[4] ==  self.gamestate[7] != 0):return engineReturnCode.WIN
+		else: return engineReturnCode.RUNNING
```
last `else` is redundant

deysumitkr reviewed (changes requested):
Looks like the design has largely fallen into place. Some minor remarks have been added.

Next, follow a consistent naming convention throughout the code base:
- [Naming convention for Python: Wikipedia](https://en.wikipedia.org/wiki/Naming_convention_(programming)#Python_and_Ruby) 
- [Python Naming Convention](https://namingconvention.org/python/)

----
General remark:
When a function has return statements under conditional blocks, make sure that there exists a default return statement without any conditions. If such default return statement does  not exist or is never reached, then it signals towards unnecessary complexity of code.

anikde committed:
Naming conventions checked
Redundent else statements were removed
Class Tictactoe has new function "get_available_pos" which assigns the available positions to a local variable named vacant_pos
** Kept the "move" the function in agent.py which returns a legal move after receiving the current gamestate as an argument

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,71 @@
+	def check_status(self):
+		if( self.check_tie() == EngineReturnCode.TIE ): return EngineReturnCode.TIE
+		if( self.check_win() == EngineReturnCode.WIN ): return EngineReturnCode.WIN
+		elif( self.check_win() == EngineReturnCode.RUNNING ): return EngineReturnCode.RUNNING
```
It appears that the return value from `self.check_win()` is returned anyway. As `self.check_win()` has only these two return values `WIN` and `RUNNING`, both of which are returned directly, there should not be any reason to increase complexity by implementing a `if else` block around it.
Just use `return self.check_win()` instead of above two lines.

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,71 @@
+		self.list_of_states.append(list(self.gamestate))
+		return self.list_of_states
+
+	def check_valid(self, players_move):
```
please, rename the function to `check_valid_input` or `check_valid_player_input` or `check_valid_player_move` etc.
just `check_valid` is not sufficiently clear - for eg: validity of the game_state can also be a reasonable assumption.

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,71 @@
+from enum import Enum, unique
+import random
+@unique
+class EngineReturnCode(Enum):
+	WIN = 0
+	RUNNING = 1
+	TIE = 2
+	VALID = 3
+	INVALID_INPUT = 4
```
This Enum class mixes two kinds constants. One for game state (win, running, tie) and other for player input (valid, invalid).
It would be nice to have two separate Enum classes instead of one.
(not absolutely required to be done in this PR itself)

anikde replied:
what is PR?

deysumitkr replied:
PR = Pull request

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -0,0 +1,71 @@
+	def get_available_pos(self):
+		self.vacant_pos =[]
+		for x in range(len(self.gamestate)):
+			if (self.gamestate[x] == 0): (self.vacant_pos).append(x)
```
It does not return anything, so there should not be a `get` in its name. There is nothing to get from this function - it only updates an internal variable. Maybe it should be called as `update_available_positions`.


This also means that function `check_tie` can now simply check if the `self.vacant_pos` is empty to determine if it is a tie situation. That is up to you. Just pointed out the possibility.

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,40 @@
+while(engine_instance.engine_status == EngineReturnCode.RUNNING):
+	state = engine_instance.gamestate
+	if (count % 2 == 1):
+		engine_instance.get_available_pos()
```
this function `get_available_pos()`, should not be called by interface and rather should be called by engine itself after every valid move/update (to update its internal variable). 
The interface should just use the variable `vacant_pos` directly.

anikde replied:
yes yes.. I was using it for debugging

deysumitkr commented on agent_random.py:
```diff
@@ -0,0 +1,16 @@
+		vacant_pos =[]
+		if(self.gamestate == [0]*9): return random.randint(0, 8)
+		for x in range(len(self.gamestate)):
+			if (self.gamestate[x] == 0): (vacant_pos).append(x)
```
agent should not be deducing the `vacant_pos` ('possible actions' in reinforcement learning terms), especially as the engine ('environment' in reinforcement learning terms) now evaluates it already.

deysumitkr commented on runengine.py:
```diff
@@ -0,0 +1,40 @@
+	state = engine_instance.gamestate
+	if (count % 2 == 1):
+		engine_instance.get_available_pos()
+		move = agent1.move(state)
```
As the possible actions (`vacant_pos`) are now available from engine itself, those positions should also be provided directly to agent (along with `state`). The agent should not be inferring possible actions from state, especially as environment is providing those.

anikde replied:
the agent now doesn't even needs to know the gamestate now, If it has the list_of_available_pos then it can directly choose one from there

deysumitkr replied:
For random selection, yes true - agent is not using state. But usually any agent in general would need the state to choose the best action from the set of possible actions. We just try to maintain that as a standard way to invoke agents in general, and not make special changes for this random agent.

deysumitkr reviewed (changes requested):
Code now looks much nicer with proper naming conventions!

Added some comments, so that the agent can use the actions provided by the environment rather than inferring possible actions by itself. (rest are minor comments)

---

Next thing to consider is code documentation.
However, as this PR is quite large, you may choose to document it in next PR.
Documentation Guide: https://realpython.com/documenting-python-code/

Code review is almost done. After the next changes are pushed, we can just discuss that if the data we are storing in json files are sufficient for our next purposes or minor changes (if at all) are required.

anikde committed:
Documented the code

deysumitkr commented on .gitignore:
```diff
@@ -193,3 +193,5 @@ tags
 # Custom rules (everything added below won't be overriden by 'Generate .gitignore File' if you use 'Update' option)
 
+### Dataset files ###
+gamelog.json
```
new line

deysumitkr commented on runengine.py:
```diff
@@ -9,26 +10,26 @@
 		print("count-->%d, move--->%d "%(count ,move))
 		# if(engine_instance.check_input == EngineReturnCode.INVALID_INPUT):
 		# 	move = agent1.move(state)
 		# 	returnn = engine_instance.updateState(move, agent1.value)
```
Are these required? print and comments? Same for below.

deysumitkr commented on runengine.py:
```diff
@@ -9,26 +10,26 @@
+
 if (engine_instance.engine_status == EngineReturnCode.WIN):
 	game_dictionary = {
 		#"times" : times,
```
what is this times? if useful, use it, otherwise remove it

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -18,11 +18,20 @@ def __init__(self):
 		self.check_input = EngineReturnCode.VALID
 
+	def update_state(self,agents_move, agent_value):
+		""" updates the state of the board with correct inputs from the player. Heart of the game.
```
~correct~

deysumitkr commented on tictactoe_engine.py:
```diff
@@ -18,11 +18,20 @@ def __init__(self):
 
+			Parameters
+			----------
+			agents_move : agent's move (numerical value between 0 - 8).
```
numerical value from the set [0, 8]

anikde committed:
Commented lines were deleted.
Corrections were made in documention.

deysumitkr reviewed (approved):
Great PR!
